src/lib.rs

```diff
@@ -82,40 +108,458 @@ pub trait Application {
+        let mut response = Response::new();
+        let res = ResponseDeliverTx::new();
+        response.set_deliver_tx(res);
+        let _ = responder.respond(response);
     }
 
     /// Consensus Connection: Called at the end of the block.  Often used to update the validator set.
-    fn end_block(&mut self, _req: &RequestEndBlock) -> ResponseEndBlock {
-        ResponseEndBlock::new()
+    fn end_block(&mut self, _req: RequestEndBlock, responder: Responder) {
+        let mut response = Response::new();
+        let res = ResponseEndBlock::new();
+        response.set_end_block(res);
+        let _ = responder.respond(response);
     }
 
     /// Consensus Connection: Commit the block with the latest state from the application.
-    fn commit(&mut self, _req: &RequestCommit) -> ResponseCommit {
-        ResponseCommit::new()
+    fn commit(&mut self, _req: RequestCommit, responder: Responder) {
+        let mut response = Response::new();
+        let res = ResponseCommit::new();
+        response.set_commit(res);
+        let _ = responder.respond(response);
     }
 }
 
+/// Represents the type of response that can be send with a Responder.
+pub enum ResponseType {
+    Info,
+    SetOption,
+    Query,
+    CheckTx,
+    InitChain,
+    BeginBlock,
+    DeliverTx,
+    EndBlock,
+    Commit,
+    Flush,
+    Echo,
+    Exception,
+}
+
+/// To be used by the application to provide a response,
+/// possibly asynchronously, for a specific request.
+pub struct Responder {
+    response_sender: Sender<(Response, oneshot::Sender<Response>)>,
+    network_sender: oneshot::Sender<Response>,
+    response_type: ResponseType,
+}
+
+impl Responder {
+    pub fn new(
+        response_sender: Sender<(Response, oneshot::Sender<Response>)>,
+        network_sender: oneshot::Sender<Response>,
+        response_type: ResponseType,
+    ) -> Self {
+        Responder {
+            response_sender,
+            network_sender: network_sender,
+            response_type,
+        }
+    }
+
+    fn check_response(&self, response: &Response) -> Result<(), ()> {
+        let has_expected = match self.response_type {
+            ResponseType::Info => response.has_info(),
+            ResponseType::SetOption => response.has_set_option(),
+            ResponseType::Query => response.has_query(),
+            ResponseType::CheckTx => response.has_check_tx(),
+            ResponseType::InitChain => response.has_init_chain(),
+            ResponseType::BeginBlock => response.has_begin_block(),
+            ResponseType::DeliverTx => response.has_deliver_tx(),
+            ResponseType::EndBlock => response.has_end_block(),
+            ResponseType::Commit => response.has_commit(),
+            ResponseType::Flush => response.has_flush(),
+            ResponseType::Echo => response.has_echo(),
+            ResponseType::Exception => response.has_exception(),
+        };
+        if !has_expected {
+            return Err(());
+        }
+        Ok(())
+    }
```

Comment:
I'm thinking whether this dynamic checking could be turned more into static checking (at least enforce it for the client / implementor) -- e.g. if there's `Responder<T: ResponseType> ...`

and the methods in `Application` would then require to be instantiated with the correct type, something like
`fn deliver_tx(&mut self, _p: RequestDeliverTx, responder: Responder<ResponseDeliverTx>) {....`

Something like this pattern: https://yoric.github.io/post/rust-typestate/

@gterzian any thoughts?

Reply:
That is an interesting idea, and I think it would require making a bunch of other stuff generic as well, for example

1. I believe that if the `Application` starts having generic methods, you can't `Box` it anymore. There are ways around that, by making `Protocol` itself generic. 
2. Unless we make `Response` itself generic, the check is going to have to be at runtime I think. And if we make `Response` generic, it will require making all sorts of other things generic as well. 

I think one might want to try to spawn one `Protocol`, or rather `Session` per "session", and type the whole thing based on the back and forth messaging you expect using for example those session types. 

You would probably still need one "master" `Protocol`, running like the current one, to handle other stuff (like query/echo and other potential "out of protocol" requests?) and to spawn typed "child" `Session<T>` or something similar. 

Probably best done in a potential follow-up.

Reply:
The current proposal looks good -- my main concern in the this proposal is that it's a little bit less type-safe than the previous one -- e.g. as a client / implementor, I may by mistake pass to `deliver_tx` a `Responder` that returns `ResponseCheckTx` (instead of `ResponseDeliverTx`). Ideally, this would be caught at compile-time rather than runtime.

Reply:
Ok, if I understand what you mean, you're concern is that the variant of request/response is a runtime thing, not compile-time. 

In the current version, each method of `Application` has a specific `Request_*` as argument and returns a specific `Response_*`. 

This version has a single `Responder`, that then does some sanity check at runtime using an enum of response types.

Loosing some "type-safety" will happen each time you use an `enum` as type. You can catch errors at compile-time, at the cost of of having to use a `struct` for each variant, versus a single `enum`. 

I could do the same by introducing a `ResponderCheckTx`, `ResponderDeliverTx`, `Responder_*`, and having a channel for each "type". 

That might be even better than using a generic approach, however you can't say it's pretty or concise code. It probably leaves plenty of room for logical errors anyway(the more insidious ones probably). 

So I think it might be a bad trade-off, you get fairly minimal "type-safety" preventing one class of programming errors at compile vs runtime, and in return you get convoluted code. 

FInally, in terms of programming error. Code that compile doesn't necessarily run as you'd expect. You can still make a mistake and use `DeliverTx` where `CheckTx` belongs, no one noticeses, and the app author just follows the compiler. That wouldn't ensure expected runtime behavior either.